### ISSUE TITLE
[WPE 2.22] video hickups after short rewind on dashjs player

### ISSUE THREAD

**Issue by przemyslaw-gorszkowski-red** [Metrological/WebKitForWayland#871]
When we play the content on dashjs player(only on version >= 2.6.6) we observe some hickups of the playback at the beginning after short rewind(2-10seconds back).

The list of links of dashjs player: https://reference.dashif.org/dash.js/

I tested on https://reference.dashif.org/dash.js/v4.1.0/samples/dash-if-reference-player/index.html (the first stream from VOD but probably it can be reproduce on any test stream available there)

The problem is also observed on reference RPi image but the result is a different. When we do the short rewind sometimes it starts to play from the seek time and then jump to some other position(and audio and video is desynchronized), sometime it stops to play video, sometime the playback has hickups.

**Comment by przemyslaw-gorszkowski-red** [comment]
Actually the problem can be also observed when we do short FF, in general the problem can be observed when we do seek to buffered range (or very close to it in case of rewind)

**Comment by przemyslaw-gorszkowski-red** [comment]
The change in dashjs player which probably "introduces" this problem on WPE: https://github.com/Dash-Industry-Forum/dash.js/pull/2342

**Comment by przemyslaw-gorszkowski-red** [comment]
The recordings of the problematic case in rewind and forward are attached in RDK ticket: https://jira.rdkcentral.com/jira/browse/WPEBROWSER-7 (they are too big to attach them here:() - all seeks there I did to 23 second of the playback

**Comment by przemyslaw-gorszkowski-red** [comment]
Some analysis base on logs([WPE2.22_video_hickups_after_short_rewind_on_dashjs_player.log.zip](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/8842240/WPE2.22_video_hickups_after_short_rewind_on_dashjs_player.log.zip)):
```
// start seek from 41.92 to 29.92
May 25 08:50:00 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:00.850060333    16    0x735c0 INFO               webkitmse MediaPlayerPrivateGStreamerMSE.cpp:233:seek: [Seek] seek attempt to {29.920000} secs
May 25 08:50:00 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:00.850135851    16    0x735c0 DEBUG              webkitmse MediaPlayerPrivateGStreamerMSE.cpp:251:seek: Seeking from {41920000000/1000000000 = 41.920000} to {29.920000} seconds

// seek was done to buffered time range
May 25 08:50:00 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:00.850240962    16    0x735c0 DEBUG              webkitmse MediaPlayerPrivateGStreamerMSE.cpp:837:isTimeBuffered: Time {29.920000} buffered? true
May 25 08:50:00 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:00.850291036    16    0x735c0 DEBUG              webkitmse MediaPlayerPrivateGStreamerMSE.cpp:837:isTimeBuffered: Time {29.920000} buffered? true

// start to remove some frames from near future (audio and video)
May 25 08:50:00 3C36E4-EOSSTB-003356051205 WPEWebProcess[11222]: 220525-06:50:00.835194 [mod=MediaSource, lvl=INFO] [tid=16] SourceBuffer::removeCodedFrames(0x7715b400) - start({59.920000}), end({72.500000})
May 25 08:50:00 3C36E4-EOSSTB-003356051205 WPEWebProcess[11222]: 220525-06:50:00.963347 [mod=MediaSource, lvl=INFO] [tid=16] SourceBuffer::removeCodedFrames(0x7715b600) - start({59.920000}), end({80.499999})

// starting the playback
May 25 08:50:01 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:01.840804073    16   0x29f1c0 INFO        brcmvideodecoder gst_brcm_video_decoder.c:1517:first_pts_pass_cb: Video decoder presentation started
May 25 08:50:01 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:02.132426628    16   0x29f1c0 DEBUG       brcmvideodecoder gst_brcm_video_decoder.c:1028:sendPositionUpdate: 30200000000
May 25 08:50:02 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:02.398326110    16   0x29f1c0 DEBUG       brcmvideodecoder gst_brcm_video_decoder.c:1028:sendPositionUpdate: 30480000000

// flushing buffers (?)
May 25 08:50:02 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:02.398991554    16    0x735c0 DEBUG              webkitmse PlaybackPipeline.cpp:333:flush: flush: trackId=V1

// restarting the playback after flush
May 25 08:50:02 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:02.819568295    16   0x29f1c0 INFO        brcmvideodecoder gst_brcm_video_decoder.c:1517:first_pts_pass_cb: Video decoder presentation started
May 25 08:50:02 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:03.098497183    16   0x29f1c0 DEBUG       brcmvideodecoder gst_brcm_video_decoder.c:1028:sendPositionUpdate: 30760000000
May 25 08:50:02 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:03.378231331    16   0x29f1c0 DEBUG       brcmvideodecoder gst_brcm_video_decoder.c:1028:sendPositionUpdate: 31040000000
May 25 08:50:03 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:03.658360664    16   0x29f1c0 DEBUG       brcmvideodecoder gst_brcm_video_decoder.c:1028:sendPositionUpdate: 31320000000

// another flushing buffers (?)
May 25 08:50:03 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:03.721364368    16    0x735c0 DEBUG              webkitmse PlaybackPipeline.cpp:333:flush: flush: trackId=V1

// restarting the playback after flush
May 25 08:50:04 3C36E4-EOSSTB-003356051205 wpe.sh[11185]: 0:03:05.019020441    16   0x29f1c0 INFO        brcmvideodecoder gst_brcm_video_decoder.c:1517:first_pts_pass_cb: Video decoder presentation started
```
IMO the flushes the buffers after starting the playback are the reason of this strange hickups.

**Comment by calvaris** [comment]
I have tried to run [this webpage](https://reference.dashif.org/dash.js/v4.1.0/samples/dash-if-reference-player/index.html) and I can't play anything on the Pi because I don't have mouse support to be able to hit the play button

**Comment by przemyslaw-gorszkowski-red** [comment]
You can do this by webinspector via: 
document.getElementsByTagName("button")[4].click() 
it will load and autoplay the selected stream.

To change the selected video stream I use:
document.querySelector("#desktop-streams > ul > li:nth-child(1) > ul > li:nth-child(2) > a").click()

To make a seek:
v = document.querySelector("video");
v.currentTime=23

**Comment by rychem** [comment]
@calvaris would you please share whether there is any progress of issue investigation?

**Comment by calvaris** [comment]
I have other priorities, it's still on my to-do.

**Comment by rychem** [comment]
Would you be able to estimate when it could investigated? 
Issue is being openned almost for 2months without initial assessment by Metrologica/Igalia.

**Comment by woutermeek** [comment]
@rychem Unfortunately we have not had the resources available to look at this issue yet. We are working hard to get to this as soon as we can. It is hard to give an actual estimate because of the priority tasks that we are currently on. Please note  that as soon as we have some one available this will be the first issue to be picked up.

**Comment by rychem** [comment]
@woutermeek in reference to last RDK Browser Working Group and my AP to precise priorities. 
We would like to ask to prioritize this issue (https://jira.rdkcentral.com/jira/browse/WPEBROWSER-7) over https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/917 (https://jira.rdkcentral.com/jira/browse/WPEBROWSER-9)

**Comment by rychem** [comment]
@woutermeek is there any progress on this ticket according to our prioritization arragement?

**Comment by woutermeek** [comment]
We've seen the priority update, unfortunately we did not have an engineer available yet to look into it.

**Comment by eocanha** [comment]
I've taken over this task, reproduced the hiccups locally by seeking from ~10s to 2s on the first VOD stream ("VOD (Static MPD) --> [DASH-IF] SegmentBase, ondemand profile":
https://dash.akamaized.net/dash264/TestCases/1a/sony/SNE_DASH_SD_CASE1A_REVISED.mpd) and analyzed the locally collected logs.

There's an initial enqueue, then the expected seek and enqueue, and then two flushes and reenqueues after that. The flushes are strange, because there's no quality change on the video. There are redundant audio and video appends (some of the video apppends are after flushes).

completeMissingMetadata() sets an offset of the duration of one sample at the begining of the first video append, keeps it for all the overwriting appends that start at 0, but removes it for a new overwriting append that doesn't start at 0. Some samples overwritten by the latter append have slightly different PTSs for the same DTS because of that difference in offset/no-offset. Still, that happens for samples far away in the future from where the hiccups are happening, so that shouldn't be the main problem.

The main cause of the hiccups might be the flush and reenqueue operations, which put some stress in the video decoder. Now I should try to analyze the dash.js code a bit to try to understand what's causing the redundant video appends. They shouldn't be needed if there's no quality change involved.

**Comment by eocanha** [comment]
I keep thinking that stress in the video decoder is a real problem (but out of our control). The decoder just can't clean itself quick enough and decode a bunch (~2 sec, the new [0, 2] ones, as 0 is the sync frame) of delayed buffers quick enough to try to produce the new next buffer (2.002) on time to be displayed in real time.

Analyzed the PTSs of the decoded buffers on Raspberry Pi (OMX) and they show what is perceived in the screen: after the seek, several flushes must happen because the output buffers are like [2, 2.75] [2, 2.043] [2, 2.043] [4.83, 7.92]. The three first ones cause stuttering (frame repetition) and the 4th one is perceived like a small stall.

Unless we can "convince" dash.js in some way not to overwrite already buffered samples, we won't be able to avoid the flushes and the described behaviour.

Experimented with changing some dash.js configuration variables, such as streaming.buffer.fastSwitchEnabled=false and streaming.buffer.flushBufferAtTrackSwitch=true without getting improvements. Added logs to check that the streaming.trackSwitchMode.video=neverReplace setting is honored (it is), but sample replacements happen anyway. Couldn't find a way to avoid sample replacements (and flushes) by now.

**Comment by rychem** [comment]
@eocanha would you please share whether there is any progress on this issue?

**Comment by eocanha** [comment]
I've been working on this all previous week (W35) and confirmed my impressions about the hiccups being caused by the stress that the flushes and reenqueues put on the video decoder. There's no way to avoid those flushes from WebKit, because WebKit directly enqueues the appended samples for playback, and if those samples are deleted or overwritten (the two cases happening here), the only reasonable thing that WebKit can do to "unenqueue" them is to flush the playback pipeline and enqueue the new versions or whatever is needed to keep playing after that, therefore stressing the decoder.

Some highlights of what I tried/found last week:
- Started to analyze dash.js code to try to understand why the quality changes (and reappends) happen.
- Checked that the problem also happens on wpe-2017 and wpe-2.28 with some slight variations probably due to timing.
- The first flush is caused by sample removal. The second one by sample overwriting (reappend).
- Tweaked some dash.js settings (bufferTimeAtTopQuality, bufferTimeAtTopQualityLongForm and bufferToKeep) without success.
- Found out a bit more about where the logic from the buffers to be removed comes from in dash.js.
- Experimented with dash.js' BufferController.getAllRangesWithSafetyFactor() and managed to avoid the deletion in a forced way. It's not valid for the general case but it's a good path to solve the issue.

This week I'm following the getAllRangesWithSafetyFactor() path. My intention is to provide a dash.js patch that can avoid deletions that cause the flush & reenqueues that stress the decoder.

**Comment by eocanha** [comment]
I've come up with a patch for dash.js (published in [this branch](https://github.com/eocanha/dash.js/commits/eocanha-4.1.0)) that solves the problem on [dash.js 4.1.0](https://github.com/Dash-Industry-Forum/dash.js/commits/v4.1.0):

https://github.com/eocanha/dash.js/commit/61192710c03144e620be5f25b482704dfbd5ffd1

**Comment by rychem** [comment]
@eocanha many thanks for your investigation and proposed fix ! Especially it was not an easy case.
Is my understanding correct that you will be planning to submit PR for dash.js master/development branch to get it on newest dash.js release?

**Comment by eocanha** [comment]
I've rebased it to the latest dash.js development branch and submitted https://github.com/Dash-Industry-Forum/dash.js/pull/4039. Let's see if they like it and decide to merge it on version 4.5.

**Comment by rychem** [comment]
Thank you Enrique very much. 
We are closing then internal ticket with reference to dash.js version 4.5 as potential fix holder and recommendation for app vendor for upgrade.

**Comment by eocanha** [comment]
Take into account that there's no guarantee yet that the change is going to be merged on dash.js. It would be advisable to suggest your app vendor to use their own patched version in the meantime.

**Comment by jacek-skiba-red** [comment]
Hello,

* on dashjs version 3.1.1 I see that player is triggering buffers removal from near future, which is causing flush (stuttering)
    * setting fastSwitchEnabled to true/false is not changing player behavior
    * on this version your above PR is valuable 
* on latest dashjs version (4.4.1) I see different behavior
    * buffers from future are never removed
    * stuttering exists, but the root cause here is that player is once again pushing to MSE the same chunk
     * similar behavior I see on chrome (the same chunk is one more time fetched), but there stuttering is not seen
 
```
// position : 616 sec
23:45:54    gst_brcm_video_decoder.c:1093:sendPositionUpdate: 616600022201

// fetch chunk 30fps_640x360_1000k, nr 158
23:45:54    OnFragmentLoadingCompleted for stream id defaultId_0 and media type video - Url: https://dash.akamaized.net/akamai/bbb_30fps/bbb_30fps_640x360_1000k/bbb_30fps_640x360_1000k_158.m4v

// apply chunk 158 to video buffer : [598, 627]
23:45:54    SourceBuffer::appendBufferInternal(0x7e654200) - append size = 553572, buffered = [{598000000/1000000 = 598.000000}..{627999999/1000000 = 627.999999}]
23:45:54    SourceBuffer::sourceBufferPrivateAppendComplete(0x7e654200) - buffered = [{598000000/1000000 = 598.000000}..{631999999/1000000 = 631.999999}]

// position : 621 sec
23:45:58    gst_brcm_video_decoder.c:1093:sendPositionUpdate: 621133333313

// seek to position : 619 sec
23:45:58    HTMLMediaElement::seek(9131ECE) {"value":619.066666646}
23:45:59    HTMLMediaElement::scheduleEvent(9131ECE) scheduleEvent: seeked 

// during seek we reject download of chunk 159
bbb_30fps_640x360_1000k_159.m4v:0,0 : Load request cancelled

// we are removing only old chunks (chunks from future are not removed)
23:45:59    SourceBuffer::remove(0x7e654200) - start(598.000000), end(599.066667) 

// we are in 620 sec
23:46:00    gst_brcm_video_decoder.c:1093:sendPositionUpdate: 620400022201

// init segment with video 640x360
23:46:00    OnFragmentLoadingCompleted for stream id defaultId_0 and media type video - Url: https://dash.akamaized.net/akamai/bbb_30fps/bbb_30fps_640x360_1000k/bbb_30fps_640x360_1000k_0.m4v
23:46:00    SourceBuffer::appendBufferInternal(0x7e654200) - append size = 706, buffered = [{600000000/1000000 = 600.000000}..{631999999/1000000 = 631.999999}]
23:46:00    MediaPlayerPrivateGStreamerMSE.cpp:927:trackDetected:track ID: V6, caps: video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)3, profile=(string)high, width=(int)640, height=(int)360, pixel-aspect-ratio=(fraction)1/1
23:46:00    SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment(0x7e654200)
23:46:00    SourceBuffer::sourceBufferPrivateAppendComplete(0x7e654200) - buffered = [{600000000/1000000 = 600.000000}..{631999999/1000000 = 631.999999}]

// position : 621 sec
23:46:00    gst_brcm_video_decoder.c:1093:sendPositionUpdate: 621733333312 

// once again we are fetching chunk 158
23:46:01    OnFragmentLoadingCompleted for stream id defaultId_0 and media type video - Url: https://dash.akamaized.net/akamai/bbb_30fps/bbb_30fps_640x360_1000k/bbb_30fps_640x360_1000k_158.m4v

// we append buffer 158 which is causing old samples reenque and flush
23:46:01    SourceBuffer::appendBufferInternal(0x7e654200) - append size = 553572, buffered = [{600000000/1000000 = 600.000000}..{631999999/1000000 = 631.999999}]
23:46:01    MediaPlayerPrivateGStreamerMSE.cpp:927:trackDetected: track ID: V6, caps: video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)3, profile=(string)high,width=(int)640, height=(int)360, pixel-aspect-ratio=(fraction)1/1, framerate=(fraction)119/4 
23:46:01    SourceBuffer::sourceBufferPrivateDidReceiveSample(0x7e654200) - removing sample({PTS({628000000/1000000 = 628.000000}), DTS({628000000/1000000 = 628.000000}),duration({33333/1000000 = 0.033333}), flags(sync), trackId(V6), presentationSize(640.000000x360.000000)})
            [...]
23:46:01    SourceBuffer::sourceBufferPrivateDidReceiveSample(0x7e654200) - removing sample({PTS({631966666/1000000 = 631.966666}), DTS({631966666/1000000 = 631.966666}),duration({33333/1000000 = 0.033333}), flags(), trackId(V6), presentationSize(640.000000x360.000000)})
23:46:02    SourceBuffer::sourceBufferPrivateDidReceiveSample(0x7e654200) removed 266641 bytes, start(630.000000), end(631.999999)
23:46:02    PlaybackPipeline.cpp:333:flush: flush: trackId=V6
23:46:02    SourceBuffer::sourceBufferPrivateAppendComplete(0x7e654200) - reenqueuing at time ({621766666646
```
* full log from reproduction: [4.4.1_stuttering.log](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/9554505/4.4.1_stuttering.log)

@eocanha, please confirm if you see similar behavior on your side on latest dashjs. 
I'm not convinced if this is a Webkit issue, however Chrome copes well with this unusual player behavior.

**Comment by eocanha** [comment]
I think the problem you comment is a different one from the one I solved. I think I experienced something similar on 4.1.0 a couple of times (update of a buffered range, unrelated to deletions, but which conflicts with already enqueued ranges and causes a flush). I thought it had been solved with fastSwitchEnabled=false.

Things go better in Chrome because it's being run on a desktop system (more powerful than a set-top-box), and it might be (I don't know) that Chrome has a different enqueuing policy and doesn't enqueue all samples as they are appended, needing for less flush and reenqueues when range reappends happen.

**Comment by jacek-skiba-red** [comment]
I'm reproducing this issue all the time on WPE2.22 an Chrome, with fastSwitchEnabled equal to false. 
Attaching devtools output:

![dashjs_4 4 1](https://user-images.githubusercontent.com/42027818/190376142-620154e9-348c-431f-a310-f2253e92addd.png)

The problem is that fastSwitchEnabled works when we want to move to higher resolution, which in not done in our situation. We are getting initialization segment with the same resolution. 

In general this is a dashjs issue, and in my opinion we can close this webkit ticket.

**Comment by eocanha** [comment]
@rychem, https://github.com/Dash-Industry-Forum/dash.js/pull/4039 has been merged into dash.js upstream as https://github.com/Dash-Industry-Forum/dash.js/commit/2935efa2b1b1f8493b5e3aab2e226be18cedd0e3. It'll be available in the next dash.js release.

The patch has gone through some changes compared with the original one I posted here. In particular, remind your app vendor that, **in order for the intended fixes to be active, they need to set the streaming.buffer.avoidCurrentTimeRangePruning=true in the player settings**.
